api/proto/teleport/discoveryconfig/v1/discoveryconfig_service.proto

```diff
@@ -0,0 +1,92 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.discoveryconfig.v1;
+
+import "google/protobuf/empty.proto";
+import "teleport/legacy/types/types.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/discoveryconfig/v1;discoveryconfigv1";
+
+// DiscoveryConfigService provides methods to manage Discovery Configs.
+//
+// Each Discovery Config has a set of matchers and a DiscoveryGroup.
+// DiscoveryServices also have a DiscoveryGroup which will be used to load all the matchers from
+// all the DiscoveryConfigs that have the same Group.
+service DiscoveryConfigService {
+  // ListDiscoveryConfigs returns a paginated list of Discovery Config resources.
+  rpc ListDiscoveryConfigs(ListDiscoveryConfigsRequest) returns (ListDiscoveryConfigsResponse);
+
+  // GetDiscoveryConfig returns the specified DiscoveryConfig resource.
+  rpc GetDiscoveryConfig(GetDiscoveryConfigRequest) returns (types.DiscoveryConfigV1);
```

Comment:
Is it possible to move DiscoveryConfigV1 to this package?

It's only used here but it's def happens in types.

Reply:
If I do that, I think I'll need to duplicate the spec fields in `api/types/discoveryconfig/...`
At least, that's what's happening with recent AccessList resources.

I would prefer to not do that.
The matchers are quite complex structs and we already have them duplicated in two places:
- for the fileconfig reader: https://github.com/gravitational/teleport/blob/42250364ef53bef1d4a913f4178300aa3db06aa2/lib/config/fileconf.go#L1837
- for the discovery service (re-used by this service)

Adding another definition of those types is a recipe for someone forgetting to add fields when evolving a particular matcher.

I tried to follow the same thing we did for the IntegrationsService: https://github.com/gravitational/teleport/pull/23553

Reply:
I am saying this because we are trying to move away from gogo and it seems a good candidate for that.
@mdwn has defined the  `FromProto` and `ToProto` helpers to translate from storage/tctl config structures and their protobuf representation for access lists

Reply:
We'll still be using gogoproto because the matchers are defined in `types.proto`

Let me see if I can make it work by moving the toplevel fields into their own package

Reply:
@tigrato Can you please take another look?